File: frontend-react/project/src/features/processing/utils/apiError.ts
/**
 * One place that turns whatever the API (or the network) throws into a sentence for the UI.
 *
 * The backend answers with:
 *  - ProblemDetails from GlobalExceptionHandler:  { title, detail, status }
 *  - ValidationProblemDetails from FluentValidation auto-validation: { title, status, errors: { Field: [msg] } }
 *  - a bare string for Unauthorized("...")
 *  - { message } from the Collection/Auth controllers
 * Duck-typed on purpose so this file does not need to import axios.
 */

interface ApiErrorBody {
  title?: string;
  detail?: string;
  message?: string;
  status?: number;
  errors?: Record<string, string[] | string>;
}

interface ErrorLike {
  message?: string;
  response?: { status?: number; data?: unknown };
  request?: unknown;
}

const MAX_VALIDATION_MESSAGES = 3;

const prettyField = (field: string): string =>
  field
    .replace(/\[(\d+)\]/g, (_m, i: string) => ` ${Number(i) + 1}`)
    .replace(/\./g, ' › ')
    .replace(/([a-z])([A-Z])/g, '$1 $2')
    .trim();

const validationMessages = (errors: NonNullable<ApiErrorBody['errors']>): string[] =>
  Object.entries(errors).flatMap(([field, msgs]) =>
    (Array.isArray(msgs) ? msgs : [msgs]).map((m) => (field ? `${prettyField(field)}: ${m}` : m)),
  );

export function getApiErrorMessage(error: unknown, fallback = 'Something went wrong. Please try again.'): string {
  if (typeof error === 'string') return error || fallback;
  if (!error || typeof error !== 'object') return fallback;

  const err = error as ErrorLike;
  const response = err.response;

  // Request left the browser but nothing came back (API down, CORS, offline).
  if (!response) {
    return err.request ? 'Cannot reach the server. Check that the API is running and try again.' : err.message || fallback;
  }

  const status = response.status;
  const data = response.data;

  if (typeof data === 'string' && data.trim()) return data;

  if (data && typeof data === 'object') {
    const body = data as ApiErrorBody;

    if (body.errors && typeof body.errors === 'object') {
      const msgs = validationMessages(body.errors);
      if (msgs.length > 0) {
        const shown = msgs.slice(0, MAX_VALIDATION_MESSAGES).join(' • ');
        return msgs.length > MAX_VALIDATION_MESSAGES ? `${shown} (+${msgs.length - MAX_VALIDATION_MESSAGES} more)` : shown;
      }
    }

    // 5xx detail is an exception message — show the friendly title instead.
    if (status !== undefined && status >= 500) return body.title || fallback;

    if (body.detail) return body.detail;
    if (body.message) return body.message;
    if (body.title) return body.title;
  }

  if (status === 401) return 'Your session has expired. Please sign in again.';
  if (status === 403) return 'You do not have permission to do this.';
  if (status === 404) return 'The requested resource was not found.';
  return fallback;
}

/** HTTP status of an API error, or undefined for network errors / non-API errors. */
export function getApiErrorStatus(error: unknown): number | undefined {
  if (!error || typeof error !== 'object') return undefined;
  return (error as ErrorLike).response?.status;
}

File: mobile_flutter/lib/core/network/api_error.dart
import 'package:dio/dio.dart';

/// Turns whatever the API (or the network) throws into one sentence for the UI.
/// A port of the web app's utils/apiError.ts, so both apps show the same messages.
///
/// The backend answers with:
///  - ProblemDetails from GlobalExceptionHandler:            { title, detail, status }
///  - ValidationProblemDetails from FluentValidation:         { title, status, errors: { Field: [msg] } }
///  - a bare string for Unauthorized("...")
///  - { message } from the Auth and Collection controllers
String apiErrorMessage(Object error, [String fallback = 'Something went wrong. Please try again.']) {
  if (error is String) return error.isEmpty ? fallback : error;
  if (error is! DioException) return fallback;

  final response = error.response;
  if (response == null) {
    return switch (error.type) {
      DioExceptionType.connectionError ||
      DioExceptionType.connectionTimeout ||
      DioExceptionType.receiveTimeout ||
      DioExceptionType.sendTimeout =>
        'Cannot reach the server. Check that the API is running and try again.',
      _ => fallback,
    };
  }

  final status = response.statusCode;
  final data = response.data;

  if (data is String && data.trim().isNotEmpty) return data;

  if (data is Map) {
    final errors = data['errors'];
    if (errors is Map && errors.isNotEmpty) {
      final messages = <String>[];
      errors.forEach((field, msgs) {
        final list = msgs is List ? msgs.map((m) => '$m') : ['$msgs'];
        for (final m in list) {
          messages.add(field is String && field.isNotEmpty ? '${_prettyField(field)}: $m' : m);
        }
      });
      if (messages.isNotEmpty) {
        const max = 3;
        final shown = messages.take(max).join(' • ');
        return messages.length > max ? '$shown (+${messages.length - max} more)' : shown;
      }
    }

    // A 5xx "detail" is an exception message — show the friendly title instead.
    if (status != null && status >= 500) return (data['title'] as String?) ?? fallback;

    for (final key in const ['detail', 'message', 'title']) {
      final value = data[key];
      if (value is String && value.isNotEmpty) return value;
    }
  }

  return switch (status) {
    401 => 'Your session has expired. Please sign in again.',
    403 => 'You do not have permission to do this.',
    404 => 'The requested resource was not found.',
    _ => fallback,
  };
}

int? apiErrorStatus(Object error) => error is DioException ? error.response?.statusCode : null;

String _prettyField(String field) => field
    .replaceAllMapped(RegExp(r'\[(\d+)\]'), (m) => ' ${int.parse(m[1]!) + 1}')
    .replaceAll('.', ' › ')
    .replaceAllMapped(RegExp(r'([a-z])([A-Z])'), (m) => '${m[1]} ${m[2]}')
    .trim();
